src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -603,6 +607,112 @@ private void EmitArray<T>(Type arrType, IEnumerable<T> items, Action<T, int> ite
             }
         }
 
+        //tries to emit opcodes for indexing an array or dictioanry
+        private (bool success, Type type) TryEmitIndexing(FunctionCallNode fcn)
+        {
+            //if we fail to emit indexing, pop the array/index off the stack.
+            var toPopCount = 0;
+            //to emit the correct msil we need to know the type of collection we are indexing.
+            var array = fcn.FormalArguments.FirstOrDefault();
+            //emit load array to stack.
+            var t = DfsTraverse(array);
+            toPopCount++;
+
+            if (t == null)
+            {
+                for (int i = 0; i < toPopCount; i++)
+                {
+                    EmitOpCode(OpCodes.Pop);
+                }
+
+                return (false, null);
+            }
+            else if (t == typeof(IDictionary))
+            {
+                if (t.IsGenericType)
+                {
+                    return (true, t.GenericTypeArguments[1]);
+                }
+                else
+                {
+                    return (true, typeof(object));
+                }
+            }
```

Comment:
Does there need to be a TODO here like `TryEmitIndexingForDictionary()`... before returning?